x-pack/legacy/plugins/maps/public/layers/styles/vector/components/color/color_map_select.js

```diff
@@ -0,0 +1,119 @@
+ */
+
+import React, { Component, Fragment } from 'react';
+
+import { EuiSuperSelect, EuiSpacer } from '@elastic/eui';
+import { ColorStopsOrdinal } from './color_stops_ordinal';
+import { COLOR_MAP_TYPE } from '../../../../../../common/constants';
+import { ColorStopsCategorical } from './color_stops_categorical';
+
+const CUSTOM_COLOR_MAP = 'CUSTOM_COLOR_MAP';
+
+export class ColorMapSelect extends Component {
+  state = {};
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    if (nextProps.customColorMap === prevState.prevPropsCustomColorMap) {
+      return null;
+    }
+
+    return {
+      prevPropsCustomColorMap: nextProps.customColorMap, // reset tracker to latest value
+      customColorMap: nextProps.customColorMap, // reset customColorMap to latest value
+    };
+  }
+
+  _onColorMapSelect = selectedValue => {
+    const useCustomColorMap = selectedValue === CUSTOM_COLOR_MAP;
+    const newProps = { useCustomColorMap };
+    newProps.color = useCustomColorMap ? null : selectedValue;
+
+    this.props.onChange({
+      ...newProps,
+      type: this.props.colorMapType,
+    });
+  };
+
+  _onCustomColorMapChange = ({ colorStops, isInvalid }) => {
+    // Manage invalid custom color map in local state
+    if (isInvalid) {
+      const newState = {
+        customColorMap: colorStops,
+      };
+      this.setState(newState);
+      return;
+    }
+
+    const newProps = {
+      useCustomColorMap: true,
+      customColorMap: colorStops,
+      type: this.props.colorMapType,
+    };
+    this.props.onChange(newProps);
+  };
+
+  render() {
+    const { color, useCustomColorMap } = this.props;
+
+    let colorStopsInput;
+    if (useCustomColorMap) {
+      if (this.props.colorMapType === COLOR_MAP_TYPE.ORDINAL) {
+        colorStopsInput = (
+          <Fragment>
+            <EuiSpacer size="s" />
+            <ColorStopsOrdinal
+              colorStops={this.state.customColorMap}
+              onChange={this._onCustomColorMapChange}
+            />
+          </Fragment>
+        );
+      } else if (this.props.colorMapType === COLOR_MAP_TYPE.CATEGORICAL) {
+        colorStopsInput = (
+          <Fragment>
+            <EuiSpacer size="s" />
+            <ColorStopsCategorical
+              colorStops={this.state.customColorMap}
+              onChange={this._onCustomColorMapChange}
+            />
+          </Fragment>
+        );
+      }
+    }
+
+    let valueOfSelected;
+    const colorMapOptionsWithCustom = [
+      {
+        value: CUSTOM_COLOR_MAP,
+        inputDisplay: this.props.customOptionLabel,
+      },
+      ...this.props.colorMapOptions,
+    ];
+
+    if (useCustomColorMap) {
+      valueOfSelected = CUSTOM_COLOR_MAP;
+    } else {
+      if (colorMapOptionsWithCustom.find(option => option.value === color)) {
+        valueOfSelected = color;
+      } else {
+        valueOfSelected = this.props.colorMapOptions[0].value;
+        this._onColorMapSelect(valueOfSelected);
```

Comment:
Took a stab at rewriting this file to avoid updating state from `render()`. This code I've written could stand a second, third, etc..... pass thinking about what's tracked using state vs. props. With that said, the error has been taken care of. Feel free to use anything that's useful and revise as needed!

```
export class ColorMapSelect extends Component {

  constructor(props) {
    super(props);
    this.state = {
      selected: this._getSelected(props),
    }
  }

  shouldComponentUpdate(nextProps, nextState, nextContext) {
    const customColorUpdates = this._getCustomColorUpdates(nextProps, this.state, nextState);
    if (nextProps.color !== this.state.prevColor ||
      nextProps.colorMapType !== this.state.prevColorMapType ||
      customColorUpdates
    ){
      this.setState({
        selected: this._getSelected(nextProps),
        prevColor: nextProps.color,
        prevColorMapType: nextProps.colorMapType,
        ...(customColorUpdates && customColorUpdates),
      });
      return true;
    }
    return false;
  }

  _getCustomColorUpdates(nextProps, prevState, nextState) {
    let newState = null;
    // TODO: Slim this down
    const internalStateUpdate =
      nextState.customColorRamp && prevState.customColorRamp &&
      !nextState.customColorRamp.every(e => prevState.customColorRamp.includes(e)) ||
      nextState.customColorPalette && prevState.customColorPalette &&
      !nextState.customColorPalette.every(e => prevState.customColorPalette.includes(e));
    if (internalStateUpdate) {
      return {};
    }

    if (nextProps.customColorRamp !== prevState.prevPropsCustomColorRamp) {
      newState = {
        ...{
          prevPropsCustomColorRamp: nextProps.customColorRamp, // reset tracker to latest value
          customColorRamp: nextProps.customColorRamp, // reset customColorRamp to latest value
        },
      };
    }

    if (nextProps.customColorPalette !== prevState.prevPropsCustomColorPalette) {
      if (!newState) {
        newState = {};
      }
      newState = {
        ...newState,
        ...{
          prevPropsCustomColorPalette: nextProps.customColorPalette, // reset tracker to latest value
          customColorPalette: nextProps.customColorPalette, // reset customColorRamp to latest value
        },
      };
    }

    return newState;
  }

  _onColorMapSelect = (selected, colorMapType = this.props.colorMapType) => {
    const useCustomColorMap = selected === CUSTOM_COLOR_MAP;
    const newProps = {};
    if (colorMapType === COLOR_MAP_TYPE.ORDINAL) {
      newProps.useCustomColorRamp = useCustomColorMap;
    } else if (colorMapType === COLOR_MAP_TYPE.CATEGORICAL) {
      newProps.useCustomColorPalette = useCustomColorMap;
    }
    newProps.color = useCustomColorMap ? null : selected;

    this.props.onChange({
      ...newProps,
      type: colorMapType,
    });
  };

  _onCustomColorMapChange = ({ colorStops, isInvalid }) => {
    // Manage invalid custom color ramp in local state
    if (isInvalid) {
      const newState = {};
      if (this.props.colorMapType === COLOR_MAP_TYPE.ORDINAL) {
        newState.customColorRamp = colorStops;
      } else if (this.props.colorMapType === COLOR_MAP_TYPE.CATEGORICAL) {
        newState.customColorPalette = colorStops;
      }
      this.setState(newState);
      return;
    }

    const newProps = {};
    if (this.props.colorMapType === COLOR_MAP_TYPE.ORDINAL) {
      newProps.useCustomColorRamp = true;
      newProps.customColorRamp = colorStops;
    } else if (this.props.colorMapType === COLOR_MAP_TYPE.CATEGORICAL) {
      newProps.useCustomColorPalette = true;
      newProps.customColorPalette = colorStops;
    }
    newProps.type = this.props.colorMapType;
    this.props.onChange(newProps);
  };

  _getColorMapOptions = ({ customOptionLabel, colorMapOptions }) => {
    return [
      {
        value: CUSTOM_COLOR_MAP,
        inputDisplay: customOptionLabel,
      },
      ...colorMapOptions,
    ];
  };

  _getSelected = ({
    color, useCustomColorRamp, useCustomColorPalette, colorMapType,
    customOptionLabel, colorMapOptions
  }) => {
    let valueOfSelected;
    const newColorMapOptions = this._getColorMapOptions({ customOptionLabel, colorMapOptions });

    const useCustom =
      colorMapType === COLOR_MAP_TYPE.ORDINAL
        ? useCustomColorRamp
        : useCustomColorPalette;
    if (useCustom) {
      valueOfSelected = CUSTOM_COLOR_MAP;
    } else {
      if (newColorMapOptions.find(option => option.value === color)) {
        valueOfSelected = color;
      } else {
        valueOfSelected = colorMapOptions[0].value;
        this._onColorMapSelect(valueOfSelected, colorMapType);
      }
    }
    return valueOfSelected;
  };

  render() {
    const { useCustomColorRamp, useCustomColorPalette } = this.props;

    let colorStopsInput;
    if (this.props.colorMapType === COLOR_MAP_TYPE.ORDINAL) {
      if (useCustomColorRamp) {
        colorStopsInput = (
          <Fragment>
            <EuiSpacer size="s" />
            <ColorStopsOrdinal
              colorStops={this.state.customColorRamp}
              onChange={this._onCustomColorMapChange}
            />
          </Fragment>
        );
      }
    } else if (this.props.colorMapType === COLOR_MAP_TYPE.CATEGORICAL) {
      if (useCustomColorPalette) {
        colorStopsInput = (
          <Fragment>
            <EuiSpacer size="s" />
            <ColorStopsCategorical
              colorStops={this.state.customColorPalette}
              onChange={this._onCustomColorMapChange}
            />
          </Fragment>
        );
      }
    }

    return (
      <Fragment>
        <EuiSuperSelect
          options={this._getColorMapOptions(this.props)}
          onChange={this._onColorMapSelect}
          valueOfSelected={this.state.selected}
          hasDividers={true}
        />
        {colorStopsInput}
      </Fragment>
    );
  }
}
```

Reply:
Thanks for suggestions. Since the above is based on older code, I can't use as-is. But it  I ended up using the lifecycle-methods as above, and pulled the callback out of render. Now it is using componentDidUpdate.